library/alloc/src/raw_vec.rs

```diff
@@ -394,6 +394,24 @@ impl<T, A: Allocator> RawVec<T, A> {
         let cap = cmp::max(self.cap * 2, required_cap);
         let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);
 
+        let alloc_size = cap.checked_mul(mem::align_of::<T>()).ok_or(CapacityOverflow)?;
+
+        // Since memory allocators tend to use power of two sized bins, find the
+        // bin size we will fall into.
+        let bin_size = usize::MAX >> (alloc_size.leading_zeros() - 1);
+
+        // Leave some room for allocators that add fixed overhead (usually one pointer-size)
+        // FIXME: maybe two pointers to be on the safe side? It could potentially be platform-dependent.
+        let aligned_alloc_size = bin_size.saturating_sub(mem::align_of::<usize>());
+
+        // Ignore cases where the value is actually smaller, due to substractiing fixed overhead
```

Comment:
This makes me notice that the current approach is basically
```rust
let new_cap = max![ len + additional, cap * 2, MIN_CAP ];
```
so I wonder if it might make sense to just make this another thing in that list, or even replacing the `cap * 2`.

`sizeof(usize).div_ceil(sizeof(T))` could be a constant like `MIN_NON_ZERO_CAP` is, so it's easy to subtract off even in length units instead of needing to switch to byte units here.

Reply:
> instead of needing to switch to byte units here

The byte units are used because what we're targeting is a byte size of (just under) a power of two. We can `const` provide "`usize` per `T`" and "`T` per `usize`", but I don't see how to calculate `(size_of::<T>() * cap).next_power_of_two() / size_of::<T>()` without scaling `cap` into byte units instead of `T` units.

Reply:
Oh, right!  Good point.